File: apps/agent-macos/Sources/EgressViewAgentCore/ObservationCSV.swift
import Foundation

/// Turns the local history into a CSV file the user can take elsewhere.
///
/// The point of the export is that the data stops being locked inside this app:
/// a spreadsheet, a notebook, or a colleague's script should be able to read it
/// without knowing anything about EgressView. So the output is plain RFC 4180
/// with no preamble, no comment lines and no merged cells -- anything clever
/// here becomes someone else's parse error.
public enum ObservationCSV {
    public static let columns = [
        "first_observed_at",
        "last_observed_at",
        "process_name",
        "bundle_id",
        "protocol",
        "local_address",
        "local_port",
        "remote_address",
        "remote_port",
        "remote_hostname",
        "bytes_in",
        "bytes_out",
        "collector",
        "confidence",
    ]

    public static func export(
        _ observations: [ConnectionObservation],
        formatter: ISO8601DateFormatter = ObservationCSV.timestampFormatter
    ) -> String {
        var lines = [columns.joined(separator: ",")]
        for observation in observations {
            // Built up statement by statement rather than as one 14-element
            // literal. The literal compiled here and defeated the type checker
            // on CI ("unable to type-check this expression in reasonable
            // time"), which is a difference in compiler version, not in taste.
            var row: [String] = []
            row.append(formatter.string(from: observation.firstObservedAt))
            row.append(formatter.string(from: observation.lastObservedAt))
            row.append(observation.processName)
            row.append(observation.bundleID ?? "")
            row.append(observation.networkProtocol.rawValue)
            row.append(observation.localAddress)
            row.append(String(observation.localPort))
            row.append(observation.remoteAddress)
            row.append(String(observation.remotePort))
            row.append(observation.remoteHostname ?? "")
            // Byte counts are measured when a connection ends, so an open
            // connection has none. Empty says "not measured"; a zero would say
            // "measured, and it was nothing", which is a different claim and a
            // false one.
            row.append(observation.bytesIn.map(String.init) ?? "")
            row.append(observation.bytesOut.map(String.init) ?? "")
            row.append(observation.collector.rawValue)
            row.append(observation.confidence.rawValue)
            lines.append(row.map(field).joined(separator: ","))
        }
        // A trailing newline: POSIX tools treat a file without one as truncated.
        return lines.joined(separator: "\n") + "\n"
    }

    public static let timestampFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    /// Quotes a value only when it has to be, and doubles any quote inside it.
    static func field(_ value: String) -> String {
        guard value.contains(where: { $0 == "," || $0 == "\"" || $0 == "\n" || $0 == "\r" })
        else { return value }
        return "\"\(value.replacingOccurrences(of: "\"", with: "\"\""))\""
    }

    /// A file name that carries the period, so exports of different windows do
    /// not overwrite each other or become indistinguishable in a folder.
    public static func suggestedFileName(from: Date, to: Date) -> String {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd-HHmm"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        return "egressview-\(formatter.string(from: from))-to-\(formatter.string(from: to)).csv"
    }
}

File: apps/agent-windows/src/EgressView.Agent.Core/ObservationCsv.cs
using System.Globalization;
using System.Text;

namespace EgressView.Agent.Core;

/// <summary>Portable RFC 4180 export of the exact local rows shown by the Windows Agent.</summary>
public static class ObservationCsv
{
    public static readonly string[] Columns =
    [
        "first_observed_at", "last_observed_at", "process_name", "process_id",
        "protocol", "local_address", "local_port", "remote_address", "remote_hostname", "remote_port", "country_code",
        "bytes_in", "bytes_out", "layer", "collector"
    ];

    public static string Export(IEnumerable<RecentFlow> observations)
    {
        var output = new StringBuilder();
        output.Append(Header);
        AppendRows(output, observations);
        return output.ToString();
    }

    public static string Header => string.Join(',', Columns) + "\r\n";

    public static string ExportRows(IEnumerable<RecentFlow> observations)
    {
        var output = new StringBuilder();
        AppendRows(output, observations);
        return output.ToString();
    }

    private static void AppendRows(StringBuilder output, IEnumerable<RecentFlow> observations)
    {
        foreach (var row in observations)
        {
            var fields = new[]
            {
                row.FirstSeen.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture),
                row.LastSeen.ToUniversalTime().ToString("O", CultureInfo.InvariantCulture),
                row.ProcessName ?? string.Empty,
                row.ProcessId.ToString(CultureInfo.InvariantCulture),
                row.Protocol,
                row.LocalAddress,
                row.LocalPort.ToString(CultureInfo.InvariantCulture),
                row.RemoteAddress,
                row.RemoteHostname ?? string.Empty,
                row.RemotePort.ToString(CultureInfo.InvariantCulture),
                row.CountryCode ?? string.Empty,
                row.BytesReceived?.ToString(CultureInfo.InvariantCulture) ?? string.Empty,
                row.BytesSent?.ToString(CultureInfo.InvariantCulture) ?? string.Empty,
                row.Layer.ToString(),
                row.Origin,
            };
            output.AppendJoin(',', fields.Select(Field)).Append("\r\n");
        }
    }

    public static string SuggestedFileName(DateTimeOffset from, DateTimeOffset to) =>
        $"egressview-{from.LocalDateTime:yyyyMMdd-HHmm}-to-{to.LocalDateTime:yyyyMMdd-HHmm}.csv";

    private static string Field(string value) =>
        value.IndexOfAny([',', '"', '\r', '\n']) < 0 ? value : $"\"{value.Replace("\"", "\"\"")}\"";
}
